dns_server/src/crawler/mod.rs

```diff
@@ -0,0 +1,567 @@
+                &self.config,
+                &address,
+                address_data,
+                AddressState::Disconnected,
+                &mut self.storage,
+                &self.command_tx,
+            );
+        }
+    }
+
+    fn handle_misbehaved(&mut self, _peer_id: N::PeerId, _error: P2pError) {
+        // Ignore all misbehave reports
+    }
+
+    fn handle_conn_event(&mut self, event: ConnectivityEvent<N>) {
+        match event {
+            ConnectivityEvent::Message { peer, message } => {
+                self.handle_conn_request(peer, message);
+            }
+            ConnectivityEvent::OutboundAccepted {
+                address,
+                peer_info,
+                receiver_address,
+            } => {
+                self.handle_outbound_accepted(address, peer_info, receiver_address);
+            }
+            ConnectivityEvent::InboundAccepted {
+                address,
+                peer_info,
+                receiver_address,
+            } => {
+                self.handle_inbound_accepted(address, peer_info, receiver_address);
+            }
+            ConnectivityEvent::ConnectionError { address, error } => {
+                self.handle_connection_error(address, error);
+            }
+            ConnectivityEvent::ConnectionClosed { peer_id } => {
+                self.handle_connection_closed(peer_id);
+            }
+            ConnectivityEvent::Misbehaved { peer_id, error } => {
+                self.handle_misbehaved(peer_id, error);
+            }
+        }
+    }
+
+    fn handle_sync_event(&mut self, _event: SyncingEvent<N>) {
+        // Ignore all sync events
+    }
+
+    fn new_address(addresses: &mut BTreeMap<N::Address, AddressData>, address: N::Address) {
+        if let Entry::Vacant(vacant) = addresses.entry(address.clone()) {
+            log::debug!("new address {} added", address.to_string());
+            vacant.insert(AddressData {
+                state: AddressState::Disconnected,
+                state_updated_at: tokio::time::Instant::now(),
+                fail_count: 0,
+                was_reachable: false,
+            });
+        }
+    }
+
+    fn get_dns_ip(address: &N::Address, p2p_port: u16) -> Option<IpAddr> {
+        // Only add nodes listening on the default port to DNS
+        match address.as_peer_address() {
+            PeerAddress::Ip4(addr)
+                if Ipv4Addr::from(addr.ip).is_global_unicast_ip() && addr.port == p2p_port =>
+            {
+                Some(Ipv4Addr::from(addr.ip).into())
+            }
+            PeerAddress::Ip6(addr)
+                if Ipv6Addr::from(addr.ip).is_global_unicast_ip() && addr.port == p2p_port =>
+            {
+                Some(Ipv6Addr::from(addr.ip).into())
+            }
+            _ => None,
+        }
+    }
+
+    /// Update address state.
+    ///
+    /// The only place where the address state can be updated.
+    fn change_address_state(
```

Comment:
I think this function should be implemented as a finite state machine. Let's talk about it.

Reply:
So as agreed, this is gonna be the next PR.